File: sdk-kotlin/src/main/kotlin/trade/decibel/sdk/errors/DecibelError.kt
package trade.decibel.sdk.errors

sealed class DecibelException(message: String, cause: Throwable? = null) : Exception(message, cause) {
    class ConfigException(message: String) : DecibelException("Configuration error: $message")
    class NetworkException(message: String, cause: Throwable? = null) : DecibelException("Network error: $message", cause)
    class ApiException(val status: Int, val statusText: String, message: String) : DecibelException("API error (status $status): $message")
    class ValidationException(message: String) : DecibelException("Validation error: $message")
    class TransactionException(val txHash: String? = null, val vmStatus: String? = null, message: String) : DecibelException("Transaction error: $message")
    class SimulationException(message: String) : DecibelException("Simulation error: $message")
    class SigningException(message: String) : DecibelException("Signing error: $message")
    class GasEstimationException(message: String) : DecibelException("Gas estimation error: $message")
    class WebSocketException(message: String) : DecibelException("WebSocket error: $message")
    class SerializationException(message: String, cause: Throwable? = null) : DecibelException("Serialization error: $message", cause)
    class TimeoutException(message: String) : DecibelException("Timeout error: $message")
}

File: sdk-swift/Sources/DecibelSDK/Errors/DecibelError.swift
import Foundation

public enum DecibelError: Error, LocalizedError {
    case config(String)
    case network(Error)
    case api(status: Int, statusText: String, message: String)
    case validation(String)
    case transaction(hash: String?, vmStatus: String?, message: String)
    case simulation(String)
    case signing(String)
    case gasEstimation(String)
    case webSocket(String)
    case serialization(String)
    case timeout(String)

    public var errorDescription: String? {
        switch self {
        case .config(let msg): return "Configuration error: \(msg)"
        case .network(let err): return "Network error: \(err.localizedDescription)"
        case .api(let status, _, let msg): return "API error (status \(status)): \(msg)"
        case .validation(let msg): return "Validation error: \(msg)"
        case .transaction(_, _, let msg): return "Transaction error: \(msg)"
        case .simulation(let msg): return "Simulation error: \(msg)"
        case .signing(let msg): return "Signing error: \(msg)"
        case .gasEstimation(let msg): return "Gas estimation error: \(msg)"
        case .webSocket(let msg): return "WebSocket error: \(msg)"
        case .serialization(let msg): return "Serialization error: \(msg)"
        case .timeout(let msg): return "Timeout error: \(msg)"
        }
    }
}
